magento/magento2
# Use mixin for common styles, to output above media queries in _extend.less

### Description (*)

This PR modifies the implementation of common styles to use a `.media-common()` mixin instead of a check for when `@media-common = true`. This has 2 advantages (respective to the "Fixed Issues" below). The first advantage is that all common styles are called from a single location in `_responsive.less`, just as the `media-width` styles are, which allows all common styles to be output above the `max-width` styles (including from `_extend.less`, even though it's imported below `_responsive.less`). This means that all media-width styles will override common styles with the same specificity, as they ought to. The second advantage is that the `.media-common()` mixin explicitly outputs all `.abs-` styles in `_extends.less`, even when compiling with Grunt, whereas before, Grunt compilation skipped over `&:extend()` calls that were nested (e.g. `.abs-action-link-button` inside `.abs-action-addto-product`). This broke styles in a standard workflow on a local site for elements such as the PDP add-to-wishlist button.
There is one slight disadvantage of this new implementation, which is that all `.abs-` styles are explicitly output (e.g. http://take.ms/jYPaE). I'm not sure how the old implementation skipped over a few of these, because it still included most of them. However, this is a nominal difference that should be well worth the advantages.
There will be one slight breaking change, which should be documented as part of the release and can easily be fixed by simply converting all instances of `& when(@media-common = true)` in custom themes to `.media-common()`.
There should not be any other breaking changes or backwards incompatibilities, for the following reasons:
1. Since this only affects styles in `_extend.less`, which the default themes do not use, it will not break specificity of any styles in the default themes.
2. For styles that are in `_extend.less` in a custom theme in a max-width breakpoint, if they are intended to override certain common styles, they must already have higher specificity, since they were output above common styles according to the old implementation, so they will not suffer from being output below the common styles in this new implementation.
3. The only possible issue would be for styles that are in `_extend.less` in a custom theme in a max-width breakpoint, if they have the same specificity as certain common styles, since this new implementation would cause them to override those common styles (unlike before). However, such styles were probably already intended to override the common styles, but since they never would have applied according to the old implementation, they were useless to start with, and I'm afraid that any issue that arises from leaving them in the code would be due to negligence.

### Fixed Issues

1. [#11098](https://github.com/magento/magento2/issues/11098): 'Mobile media query inside _extend.less, wrong order in styles-m.css and styles-l.css'
    * (The issue was actually closed due to the suggestion of importing `_extend.less` before `_responsive.less`, which modifies intended behavior. However, this PR fixes the issue without doing that.)
2. [#7231](https://github.com/magento/magento2/issues/7231): 'Different results from compiling less with Grunt or with "server side compilation"'
    * (This PR effectively fixes the issue, since there are no longer visual differences on the frontend between Grunt compilation and "server-side compilation", but it's worth noting that there are still slight differences in CSS when compared line-by-line.)

### Manual testing scenarios (*)

1. Create a custom theme that extends Magento/blank or Magento/luma
2. Create an `_extend.less` file in the custom theme
3. Add a min-width `.media-width` group, a max-width `.media-width` group, and a `.media-common` group
4. Add various styles to each group with the same selectors and specificity
5. Recompile CSS with either Grunt or server-side compilation
6. Confirm that all common styles in `styles-m.css` (including those from `_extend.less`) are output above all media queries, so that even the max-width styles override the common styles.

### Contribution checklist (*)
